### PR TITLE
Generated name override

### PR DESCRIPTION
**Pull request by justsmth**
Addresses issue: https://github.com/rust-lang/rust-bindgen/issues/1375

* Adds `generated_name_override` callback to the `ParseCallbacks` trait.
* Callback is invoked for each function name (after other name-related logic).
* Added test to verify.

This PR is a cleaned up version of previous: https://github.com/rust-lang/rust-bindgen/pull/2214

### CODE REVIEW

File: tests/parse_callbacks/mod.rs [hunk 1]
```diff
@@ -1,5 +1,29 @@
...
+        RemoveFunctionPrefixParseCallback {
+            remove_function_prefix: Some(prefix.to_string()),
+        }
+    }
+}
+
+impl ParseCallbacks for RemoveFunctionPrefixParseCallback {
+    fn generated_name_override(&self, function_name: &str) -> Option<String> {
+        if let Some(prefix) = &self.remove_function_prefix {
+            if function_name.starts_with(prefix) {
```
Comment (pvdrz) [inline tests/parse_callbacks/mod.rs:19]: maybe you can use [`String::strip_prefix`](https://doc.rust-lang.org/std/string/struct.String.html#method.strip_prefix) instead?

**Review by pvdrz — APPROVED** [review]

**Comment by pvdrz** [comment]
@justsmth I have one question. What happens if you use this new callback to set all the function names to the same value?

**Comment by justsmth** [comment]
> @justsmth I have one question. What happens if you use this new callback to set all the function names to the same value?

I've not tried that, but at best the bindings generated would not compile due to name conflicts. We could potentially detect/prevent such a mistake within rust-bindgen. However, I'm not sure doing so would be worth the added complexity.

File: src/ir/function.rs [hunk 2]
```diff
@@ -635,6 +635,12 @@ impl ClangSubItemParser for Function {
             // but seems easy enough to handle it here.
             name.push_str("_destructor");
         }
+        if let Some(callbacks) = context.parse_callbacks() {
+            if let Some(nm) = callbacks.generated_name_override(&name) {
+                name = nm;
+            }
+        }
+        assert!(!name.is_empty(), "Empty function name.");
```
Comment (pvdrz) [inline src/ir/function.rs:643]: Not a blocker for this PR but in the future we might want to change `ParseResult` to float up this kind of errors nicely.

**Review by pvdrz — APPROVED** [review]

**Pull request merged**